### DISCUSSION TITLE
MIDI device mapping

### DISCUSSION THREAD

**Discussion by webprofusion-chrisc** [soundshed/soundshed-app#17] [Ideas]
It would be useful for some users to be able to connect a midi device to their computer then map commands from that to commands in the app, to be relayed to the amp. Any thoughts & ideas?

**Comment by jydmnd** [comment]
Definitely something I would be interested in! I started tinkering with the idea and was able to make a small Node app that was able to change the channel, but it is very much a work in progress.

**Reply by webprofusion-chrisc** [reply]
It would be great if we could figure out how to do this with a standard web api (which I think there is). The node stuff is good for experiments but things like the bluetooth support was very patchy. That's why I ported the web version of the app.

**Reply by webprofusion-chrisc** [reply]
Got web midi working, just trying to get the electron version to behave as well. Idea will be to "train" selecting each preset by sending your midi event from your selected controller (i.e. a keyboard etc).

**Comment by webprofusion-chrisc** [comment]
The first draft of a feature for input mappings has now been added. This allows midi and (computer) keyboard inputs to be mapped to amp channel selections. Currently keys 1-4 are mapped to amp channels, Midi notes C3 to F3 are also mapped to channels. A learn function (to train different keys/inputs) is still planned.

**Reply by genmce** [reply]
I would love to be able to map control knobs and on/off switches to midi controller, like the bcr2000.

Using and autohotkey script (pc only) with the electron version, I was able to map the on/off controls of each element: Gate, comp, drive, amp, mod, delay, reverb to a computer hotkey. However, they are position dependent, and require autohotkey to be installed.  So if one moves or resizes the gui out of the state when it opens, the coordinates change and that breaks these switches.

Code below, not sure how the code tags work here, preview does not seem to change the look of it - or there are other parts of the script that cause the code tag to work weird.
I am removing the code tags

------------------------------------------ script begins here -----------------------------
/*
This is an autohokey script.
To use this script you must have autohotkey installed

Description: This script will send on/off clicks to the electron version of soundshed a control app for the spark 40 amp
On/Off clicks defined below
*Note - the location of the controls is hard coded,  so if you resize the soundshed app,  the locations won't match****
Sorry,  have not figured that out yet.

*/

#SingleInstance, force
#NoEnv  			
#MaxHotkeysPerInterval, 200
SendMode Input  		    	; Recommended for new scripts due to its superior speed and reliability.
SetTitleMatchMode, 2
SetWorkingDir %A_ScriptDir%  	;  Ensures a consistent starting directory.
CoordMode, Mouse, Relative

;~ -------------------------------------------------------------------------------------------------
;~ ---------------------------       AHK script for SoundShed app       ----------------------------
;~ --------------------------- *** SoundShed must be active window ***  ----------------------------
;~ ---------------------------        to toggle on/off switches         ----------------------------
;~ ---------------------------                 g = gate                 ----------------------------
;~ ---------------------------                 c = comp                 ----------------------------
;~ ---------------------------                d = drive                 ----------------------------
;~ ---------------------------                 a = amp                  ----------------------------
;~ ---------------------------                 m = mod                  ----------------------------
;~ ---------------------------                l = delay                 ----------------------------
;~ ---------------------------                 v = verb                 ----------------------------
;~ -------------------------------------------------------------------------------------------------

; with app not web browser!!!!!!


#ifwinactive Soundshed ; If this window is active the following keys will toggle switches on Soundshed app/spark
; hotkey format 
; key::Function(x cordinate,  y coordinate)
; use windows spy to determine coordinates - this is what they were on my screen 
	
	
	g::SparkSwitch(150, 505) 		; turn gate on/off switches are 175px apart on x coords
	c::SparkSwitch(330, 505) 		; turn comp on/off
	d::SparkSwitch(510, 505 )		; Drive on/off
	a::SparkSwitch(685, 505 )		; amp on/off
	m::SparkSwitch(860, 505 )	; mod on/off
	l::SparkSwitch(1035, 505 ) 	; DELAY on/off
	v::SparkSwitch(1210, 505 ) 	; Reverb on/off

SparkSwitch(xShed, yShed) ; function to send mouseclick to control without moving mouse
{			
			SetControlDelay, -1
			ControlClick, x%xShed% y%yShed%,Soundshed,,left,1, NA ; send {click, 150, 505}		
			return
}
#IfWinActive
---------------------script ends here----------------------

**Reply by genmce** [reply]
I don't understand why only part of the code is in code view... sorry...

**Reply by fefillo** [reply]
genmce, this looks great! i have not tried this yet, buy do you know if i need to adjust your parameters based on my screen resolution? and if so... what's the best way to find out screen positions?

**Reply by fefillo** [reply]
chris, do you have any details on the midi message soundshed is looking for? all I see on the setting tab is "48, 50, 52..." but on my midi pedal I have a lot more settings to choose from: 
MIDI Type: Program Change: Midi channel (1-16), PC number (0-127)
MIDI Type: Control Change: Midi channel (1-16), CC number (0-127) and CC value (0-127)


I tried various combinations Midi Channel, CC number and values (1,48,0 / 1,48,1 / 1,0,48 / etc) but could not get Soundshed to recognize them. The keyboard shortcuts are working great.

Thanks, Fefo
